Update Middleware documentation (#37992)
vercel/next.js: docs/api-routes/edge-api-routes.md

---
description: Edge API Routes enable you to build high performance APIs directly inside your Next.js application.
---

# Edge API Routes (Beta)

> Edge API Routes enable you to build high performance APIs with Next.js. By using the [Edge Runtime](/docs/api-reference/edge-runtime.md), they are faster and more secure than Node.js based API Routes. This performance improvement does come with [constraints](/docs/api-reference/edge-runtime.md#unsupported-apis), like not having access to native Node.js APIs. Instead, Edge API Routes are built on standard Web APIs.

Reviewer: ```suggestion
Edge API Routes enable you to build high performance APIs with Next.js. By using the [Edge Runtime](/docs/api-reference/edge-runtime.md), they are faster and more secure than Node.js-based API Routes. This performance improvement does come with [constraints](/docs/api-reference/edge-runtime.md#unsupported-apis), like not having access to native Node.js APIs. Instead, Edge API Routes are built on standard Web APIs.
```

Reviewer: Also, a blanket claim that they are faster and more secure feels a little bold to me - perhaps "are often faster and more secure" would be a little less absolutist?

> Edge API Routes enable you to build high performance APIs with Next.js. By using the [Edge Runtime](/docs/api-reference/edge-runtime.md), they are faster and more secure than Node.js based API Routes. This performance improvement does come with [constraints](/docs/api-reference/edge-runtime.md#unsupported-apis), like not having access to native Node.js APIs. Instead, Edge API Routes are built on standard Web APIs.

Reviewer: ```suggestion
Edge API Routes enable you to build high performance APIs with Next.js. Using the [Edge Runtime](/docs/api-reference/edge-runtime.md), they are often faster than Node.js-based API Routes. This performance improvement does come with [constraints](/docs/api-reference/edge-runtime.md#unsupported-apis), like not having access to native Node.js APIs. Instead, Edge API Routes are built on standard Web APIs.
```

```typescript
import type { NextRequest } from 'next/server'

export const config = {
  runtime: 'experimental-edge',
}

export default async function handler(req: NextRequest) {
  return new Response(
    JSON.stringify({
      name: 'Jim Halpert',
    }),
    {
      status: 200,
      headers: {
        'content-type': 'application/json',
      },
    }
  )
}
```

### Cache-Control

```typescript
import type { NextRequest } from 'next/server'

export const config = {
  runtime: 'experimental-edge',
}

export default async function handler(req: NextRequest) {
  return new Response(
    JSON.stringify({
      name: 'Jim Halpert',
    }),
    {
      status: 200,
      headers: {
        'content-type': 'application/json',
        'cache-control': 'public, s-maxage=1200, stale-while-revalidate=600',
      },
    }
  )
}
```

### Query Parameters

```typescript
import type { NextRequest } from 'next/server'

export const config = {
  runtime: 'experimental-edge',
}

export default async function handler(req: NextRequest) {
  const { searchParams } = new URL(req.url)
  const email = searchParams.get('email')
  return new Response(email)
}
```

## Differences between API Routes

> Edge API Routes are built on the [Edge Runtime](https://edge-runtime.vercel.app/) – they have the same API signature as [Edge Middleware](/docs/advanced-features/middleware). However, Edge API Routes can [stream responses](https://edge-runtime.vercel.app/features/available-apis#web-stream-apis) from the server and run _after_ cached files (e.g. HTML, CSS, JavaScript) have been accessed.

Reviewer 2: I feel like this section isn't too beginner friendly. Many web devs won’t know what “streaming responses from the server” means or what its significance is. Many also won’t know what Edge Middleware is, and by saying “they have the same signature as Edge Middleware”, we're requiring the user to click and read another page to understand what Edge API routes are.

> The [Edge Runtime](https://edge-runtime.vercel.app/) does impose some contstraints to enable high performance and security:

Reviewer 2: I like this sentence a lot and would love to see a couple sentences on the top use cases for edge api routes. Even after reading this page, I'm still lost as to when I should use them over traditional api routes.
